### PR TITLE
Add CRUD for customer addresses

### PR DESCRIPTION
**Pull request by ben-mig**
supports DAB-485

[BigCommerce Customer Addresses](https://developer.bigcommerce.com/docs/rest-management/customers/customer-addresses#get-all-customer-addresses)

We can now:
- Get all customer addresses
- Get a single address
- Create 
- Update 
- Delete

### CODE REVIEW

File: bigc/resources/customers.py [hunk 1]
```diff
@@ -81,6 +85,40 @@ def update_form_field(self, customer_id: int, field_name: str, value: Any) -> di
         }]
         return self._api.v3.put('/customers/form-field-values', json=payload)[0]
 
-    def delete(self, customer_id: int) -> None:
-        """Delete a specific customer by its ID"""
-        self._api.v3.delete(f'/customers?id:in={customer_id}')
+    def get_addresses(self, customer_id: int) -> dict:
+        """Get all addresses, filtering for a single customer"""
+        return self._api.v3.get(f'/customers/addresses?customer_id:in={customer_id}')
```
Comment (r-thomson) [inline bigc/resources/customers.py:90]: * Following our naming conventions, this should be called `all_addresses`
* The return type should be `Iterator[dict]`
* You should be using `_api.v3.get_many()` to handle pagination

File: bigc/resources/customers.py [hunk 2]
```diff
@@ -81,6 +85,40 @@ def update_form_field(self, customer_id: int, field_name: str, value: Any) -> di
...
+            address_key: address_value
+            for address_key, address_value in address.items()
+            if address_key in valid_address_params
+        } for address in address_data]
+
+        return self._api.v3.put('/customers/addresses', json=validated_address_data)
+
+    def delete_address(self, addresses_to_delete: list) -> dict:
+        """Delete addresses by their IDs"""
+        return self._api.v3.delete(f"/customers/addresses?id:in={','.join(str(i) for i in addresses_to_delete)}")
```
Comment (r-thomson) [inline bigc/resources/customers.py:124]: We do not currently have a pattern for bulk operations in this library. These functions should be reworked to only operate on one address at a time— see the customer methods in this file for reference.

**Review by r-thomson — CHANGES_REQUESTED** [review]

File: bigc/resources/customers.py [hunk 3]
```diff
@@ -81,6 +85,40 @@ def update_form_field(self, customer_id: int, field_name: str, value: Any) -> di
...
+        valid_address_params = ['customer_id', 'first_name', 'last_name', 'company', 'phone', 'address_type',
+                                'address1', 'address2', 'city', 'country_code', 'state_or_province', 'postal_code']
+
+        validated_address_data = [{
+            address_key: address_value
+            for address_key, address_value in address_data.items()
+            if address_key in valid_address_params
+        }]
+
+        return self._api.v3.post('/customers/addresses', json=validated_address_data)
```
Comment (r-thomson) [inline bigc/resources/customers.py:107]: I don't really know what the point of this validation is. If the user provides a key that BigCommerce doesn't expect, then we should just let BigCommerce handle that. We don't want to put our own validation in the middle.

Here's how I'd expect this to work, based on how we've done things in this library so far:

```suggestion
    def create_address(self, customer_id: int, **kwargs) -> dict:
        """Add an address to the customer's address book"""
        payload = {
            'customer_id': customer_id,
            **kwargs,
        }

        return self._api.v3.post('/customers/addresses', json=payload)
```

File: bigc/resources/customers.py [hunk 4]
```diff
@@ -81,6 +85,40 @@ def update_form_field(self, customer_id: int, field_name: str, value: Any) -> di
...
+            address_key: address_value
+            for address_key, address_value in address_data.items()
+            if address_key in valid_address_params
+        }]
+
+        return self._api.v3.put('/customers/addresses', json=validated_address_data)
+
+    def delete_address(self, addresses_to_delete: list) -> dict:
+        """Delete an address by its ID"""
+        return self._api.v3.delete(f"/customers/addresses?id:in={','.join(str(i) for i in addresses_to_delete)}")
```
Comment (r-thomson) [inline bigc/resources/customers.py:124]: This is still a bulk delete when it should just be deleting a single address.

```suggestion
    def delete_address(self, address_id: int) -> None:
        """Delete an address from a customer's address book"""
        self._api.v3.delete(f'/customers/addresses?id:in={address_id}')
```

**Review by r-thomson — CHANGES_REQUESTED** [review]

File: bigc/resources/customers.py (hunk 3 above)
Comment (ben-mig) [reply bigc/resources/customers.py:107]: We're not "throwing our own" error, moreso just cleaning the data that gets passed to Bigcommerce but I see your point

**Review by r-thomson — APPROVED** [review]

**Review by fieryprophet — APPROVED** [review]

**Pull request merged**